### PR TITLE
Modification status

### PR DESCRIPTION
**Pull request by jswift-stfc**
Very basic modification status feature.

- Main window uses current input filename as title.
- Asterisk is placed after the name is modification is detected.
- Handle loading/initialising widgets by using semaphores.
- File deemed as modified if a value is changed somewhere (even if it is changed back to its original value afterwards).
- Essentially `setModified` is called in (almost) all of the slots for the data entry widgets across the application.
- In future it might be better to keep a copy of the initial `GudrunFile` object, and compare the string representation of that against the current one, during the `setModified` call - if the strings are equal - then it has not been modified, so this can handle reverts.

### CODE REVIEW

**Comment by trisyoungs** [comment]
Getting a segfault which is making testing difficult!  If I click on almost anything in the main `QTreeView` I get:
```
  File "/home/tris/src/GudPy/src/gui/widgets/container_widget.py", line 142, in handleGeometryChanged
    self.geometryInfoStack_.setCurrentIndex(self.container.geometry.value)
AttributeError: 'ContainerWidget' object has no attribute 'geometryInfoStack_'
```

**Comment by jswift-stfc** [comment]
> Getting a segfault which is making testing difficult! If I click on almost anything in the main `QTreeView` I get:
> 
> ```
>   File "/home/tris/src/GudPy/src/gui/widgets/container_widget.py", line 142, in handleGeometryChanged
>     self.geometryInfoStack_.setCurrentIndex(self.container.geometry.value)
> AttributeError: 'ContainerWidget' object has no attribute 'geometryInfoStack_'
> ```

Oops, that must have got broken in the rebase I made.. Fixed now.

**Comment by trisyoungs** [comment]
Saving the file again ("Save" or "Save As..") doesn't reset the modification status...

File: src/gui/widgets/beam_widget.py [hunk 1]
```diff
@@ -330,14 +366,17 @@ def handleRemoveBeamProfileValue(self):
...
 
     def initComponents(self):
         """
         Loads the UI file for the BeamWidget object,
         and then populates the child widgets with their
         corresponding data from the attributes of the Beam object.
         """
-
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/beam_widget.py:379]: A more descriptive name would be nice here.  `widgetsRefreshing` or similar?

File: src/gui/widgets/container_widget.py [hunk 2]
```diff
@@ -402,13 +436,16 @@ def handleRemoveElement(self):
...
+            self.parent.setModified()
 
     def initComponents(self):
         """
         Populates the child widgets with their
         corresponding data from the attributes of the Container object.
         """
-
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/container_widget.py:448]: Same here.

File: src/gui/widgets/instrument_widget.py [hunk 3]
```diff
@@ -624,13 +689,17 @@ def handleRemoveGroupingParameter(self):
...
+            self.parent.setModified()
 
     def initComponents(self):
         """
         Loads the UI file for the InstrumentWidget object,
         and then populates the child widgets with their
         corresponding data from the attributes of the Instrument object.
         """
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/instrument_widget.py:702]: And here!

File: src/gui/widgets/main_window.py [hunk 4]
```diff
@@ -266,3 +284,11 @@ def runGudrun_(self):
                 self, "GudPy Error",
                 "Couldn't find gudrun_dcs binary and/or purge_det binary."
             )
+
+    def setModified(self):
+        from inspect import currentframe, getouterframes
+        caller = getouterframes(currentframe(), 2)[1][3]
+        print(caller)
```
Comment (trisyoungs) [inline src/gui/widgets/main_window.py]: Debug output.

File: src/gui/widgets/normalisation_widget.py [hunk 5]
```diff
@@ -511,13 +555,17 @@ def handleRemoveElement(self):
...
+            self.parent.setModified()
 
     def initComponents(self):
         """
         Loads the UI file for the NormalisationWidget object,
         and then populates the child widgets with their
         corresponding data from the attributes of the Normalisation object.
         """
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/normalisation_widget.py:568]: And here!

File: src/gui/widgets/sample_background_widget.py [hunk 6]
```diff
@@ -49,6 +48,8 @@ def setSampleBackground(self, sampleBackground):
             SampleBackground object belonging to the GudrunFile.
         """
         self.sampleBackground = sampleBackground
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/sample_background_widget.py:52]: You get the idea...

File: src/gui/widgets/sample_widget.py [hunk 7]
```diff
@@ -639,13 +707,16 @@ def handleRemoveResonanceValue(self):
...
+            self.parent.setModified()
 
     def initComponents(self):
         """
         Populates the child widgets with their
         corresponding data from the attributes of the Sample object.
         """
-
+        # Acquire the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/sample_widget.py:719]: Last one!

File: src/gui/widgets/instrument_widget.py [hunk 8]
```diff
@@ -945,3 +1014,5 @@ def initComponents(self):
         self.removeGroupingParameterButton.clicked.connect(
             self.handleRemoveGroupingParameter
         )
+        # Release the lock
+        self.semaphore = True
```
Comment (trisyoungs) [inline src/gui/widgets/instrument_widget.py:1018]:
```suggestion
        self.semaphore = False
```

**Review by trisyoungs — APPROVED** [review]
:+1: Some name change suggestions and a small bug to fix, otherwise OK.

**Pull request merged**